Engine/DataFeeds/ApiDataProvider.cs

```diff
@@ -51,40 +52,70 @@ public ApiDataProvider()
         /// <returns>A <see cref="Stream"/> of the data requested</returns>
         public Stream Fetch(string key)
         {
-            if (File.Exists(key))
-            {
-                return new FileStream(key, FileMode.Open, FileAccess.Read, FileShare.Read);
-            }
-
-            // If the file cannot be found on disc, attempt to retrieve it from the API
             Symbol symbol;
             DateTime date;
             Resolution resolution;
 
+            // Fetch the details of this data request
             if (LeanData.TryParsePath(key, out symbol, out date, out resolution))
             {
-                Log.Trace("ApiDataProvider.Fetch(): Attempting to get data from QuantConnect.com's data library for symbol({0}), resolution({1}) and date({2}).",
+                if (!File.Exists(key) || IsOutOfDate(resolution, key))
+                {
+                    return DownloadData(key, symbol, date, resolution);
+                }
+
+                // Use the file already on the disk
+                return new FileStream(key, FileMode.Open, FileAccess.Read, FileShare.Read);
+            }
+
+            Log.Error("ApiDataProvider.Fetch(): failed to parse key {0}", key);
+            return null;
+        }
+
+        /// <summary>
+        /// Determine if the file is out of date based on configuration and needs to be updated
+        /// </summary>
+        /// <param name="resolution">Data resolution</param>
+        /// <param name="filepath">Path to the file</param>
+        /// <returns>True if the file is out of date</returns>
+        /// <remarks>Files are only "out of date" for Hourly/Daily data because this data is stored all in one file</remarks>
+        private bool IsOutOfDate(Resolution resolution, string filepath)
+        {
+            return resolution >= Resolution.Hour &&
+                (DateTime.Now - TimeSpan.FromDays(_downloadPeriod)) > File.GetLastWriteTime(filepath);
```

Comment:
ah nice! 👍 think we should be able to make couple of simple unit test cases for this guy